prover/crates/bin/prover_autoscaler/src/global/scaler.rs

```diff
@@ -0,0 +1,344 @@
+        let mut gpu_pools: Vec<GPUPool> = clusters
+            .clusters
+            .values()
+            .flat_map(|c| self.convert2gpu_pool(namespace, c))
+            .collect();
+
+        gpu_pools.sort_by(|a, b| {
+            a.gpu
+                .cmp(&b.gpu) // Sort by GPU first.
+                .then(
+                    a.sum_by_pod_status(PodStatus::NeedToMove)
+                        .cmp(&b.sum_by_pod_status(PodStatus::NeedToMove)),
+                ) // Sort by need to evict.
+                .then(
+                    a.sum_by_pod_status(PodStatus::LongPending)
+                        .cmp(&b.sum_by_pod_status(PodStatus::LongPending)),
+                ) // Sort by long Pending pods.
+                .then(a.preemtions.cmp(&b.preemtions)) // Sort by preemtions in the cluster.
+                .then(
+                    self.cluster_priorities
+                        .get(&a.name)
+                        .unwrap_or(&1000)
+                        .cmp(self.cluster_priorities.get(&b.name).unwrap_or(&1000)),
+                ) // Sort by priority.
+                .then(b.max_pool_size.cmp(&a.max_pool_size)) // Reverse sort by cluster size.
+        });
+
+        gpu_pools
+    }
+
+    fn speed(&self, gpu: GPU) -> u64 {
+        self.prover_speed
+            .get(&gpu)
+            .cloned()
+            .unwrap_or(DEFAULT_SPEED)
+    }
+
+    fn provers_to_speed(&self, gpu: GPU, n: u32) -> u64 {
+        self.speed(gpu) * n as u64
+    }
+
+    fn normilize_queue(&self, gpu: GPU, q: u64) -> u64 {
+        let speed = self.speed(gpu);
+        // Divide and round up if there's any remainder.
+        (q + speed - 1) / speed * speed
+    }
+
+    fn run(&self, namespace: &String, q: u64, clusters: &Clusters) -> HashMap<GPUPoolKey, u32> {
+        let sc = self.sorted_clusters(namespace, clusters);
+        dbg!("run, sorted_clusters", namespace, &sc);
+
+        let mut total: i64 = 0;
+        let mut provers: HashMap<GPUPoolKey, u32> = HashMap::new();
+        for c in &sc {
+            for (status, p) in &c.provers {
+                match status {
+                    PodStatus::Running | PodStatus::Pending => {
+                        total += self.provers_to_speed(c.gpu, *p) as i64;
+                        provers
+                            .entry(c.to_key())
+                            .and_modify(|x| *x += p)
+                            .or_insert(*p);
+                    }
+                    _ => (), // Ignore LongPending as not running here.
+                }
+            }
+        }
+
+        // Remove unneeded.
+        if (total as u64) > self.normilize_queue(GPU::L4, q) {
+            for c in sc.iter().rev() {
+                let mut excess_queue = total as u64 - self.normilize_queue(c.gpu, q);
+                let mut excess_provers = (excess_queue / self.speed(c.gpu)) as u32;
+                let p = provers.entry(c.to_key()).or_default();
+                if *p < excess_provers {
```

Comment:
I'm not sure you need to deref the value here, AFAIK after `or_default` it returns mutable reference for the entry, so you can modify it right away

Reply:
I tried, but you can't. I think deref is required here.

Reply:
sounds strange, but fine